### PR TITLE
Security V1.3.1

### PR DESCRIPTION
**Pull request by JordiMarias**
This pull request adds the functionality of the security V1.3.1; as specified in the following document:
https://www.etsi.org/deliver/etsi_ts/103000_103099/103097/01.03.01_60/ts_103097v010301p.pdf

I've opened this pull request to get the code already developed to be reviewed, so I'm sure we are on the same page. Although the V1.3.1 it's already developed and working, I'm planning to update the Certify tool and add testing in the future. To run this new security layer it can be done by compiling "socktap" and running it with the following command and the certificates below:
./socktap -p static --certificate 3_Root_CA_Cert.bin --certificate-key 3_Root_CA_private.pem --security certs --security-version 3

To ensure retro compatibility I've used boost::variant on both the Certificate class and the SecuredMessage class. Which are split into SecuredMessageV2 and SecuredMessageV3 and used equally with SecuredMessageVariant. Which can be fundamentally used the same way the old SecuredMessage object could. Exactly the same applies to the CertificateVariant object.

Finally both the "verify service" and the "sign service" are under an interface and can be used indistinctly and adapt themselves to the 
the version of the security used.

I'll look forward to and answer any review.

P.D: I've developed this for a project and still have some months ahead to further develop features of the security layer. My intention is NOT to directly merge my code is to start a review and improvement process until an optimal point of development is reached.

Example Certificate:

**Certificate**: https://drive.google.com/file/d/1nA4nwp5ItymthmwnrTeVWYBhlSaR3Az2/view?usp=sharing
**Private** **Key**: https://drive.google.com/file/d/15qBeUI1zDyZKk76VDQ7iuAUkrbkV-Pdh/view?usp=sharing

### CODE REVIEW

**Review by riebl — CHANGES_REQUESTED** [review]
Thanks for your effort and sorry for my delayed review. Looks quite good overall, I have just some minor change requests. See the inline comments for details.

File: vanetza/common/archives.hpp [hunk 1]
```diff
@@ -34,6 +34,8 @@ class InputArchive
         return *this;
     }
 
+    size_t get_available();
```
[the remark this answers is not in the document]
Comment (JordiMarias) [reply vanetza/common/archives.hpp:37]: This addition is used at the "size_t deserialize(InputArchive& ar, SecuredMessageV3& message)" to get the bytebuffer of the SecuredMessageV3 to then be deserialized into the asn1c object.

File: vanetza/security/certificate.hpp [hunk 2]
```diff
@@ -214,6 +216,221 @@ boost::optional<ecdsa256::PublicKey> get_public_key(const Certificate&, Backend&
...
+         * \brief The version of the standard of the certificate
+         * \return version (3)
+         */
+        uint8_t version() const { return 3; }
+        /**
+         * \brief Conversor between Asn1c object to vanetza object (EccPoint)
+         * \param curve_point asn1c curve_point
+         * \return Vanetza EccPoint
+         */
+        static EccPoint EccP256CurvePoint_to_EccPoint(const EccP256CurvePoint_t& curve_point);
```
[the remark this answers is not in the document]
Comment (JordiMarias) [reply vanetza/security/certificate.hpp]: I've moved all of these functions to the vanetza::asn1 namespace in the vanetza/asn1/utils.hpp file.

File: vanetza/security/tests/certificate.cpp [hunk 3]
```diff
@@ -63,3 +77,35 @@ TEST(Certificate, WebValidator_AuthorizationTicket1_v2)
...
+    std::string hex_string = "0380810019000f816f72746f6163742e73652e746f63006d00000000101c185b0086012380020202806e01010280700202803801010181e00301ff0100c08280826ceb314218c4c47df101b0e0526b273d69c54962e0feddd4b3c56a9efc944980806c82318218ebc442f1c4b07d520127e0696b493de0c5dd62b3fe6ad4fcc5499e809465807d17793757ad5e284a145aa8b239b58b17c7997e3dd63b257c48229df30276a832935d2be3d9634802188b18538ddf34b3895123ebd5d2d107c71203863f0035";
```
[the remark this answers is not in the document]
Comment (JordiMarias) [reply vanetza/security/tests/certificate.cpp]: For the project, we are partnering with another company, which as far as I'm concerned is interpreting the standard to then create the certificates.
I'm also trying to get a commercial V2x software stack to do a few interoperability tests. But I'm still waiting to find one stack that fully implements the 1.3.1 stack.

**Comment by JordiMarias** [comment]
Hi, again Riebl! I know haven't been active in this PR for quite a long time, but I'd like to contribute to the project and merge it. 
I've finally reviewed all the comments again and it seems now that everything is in order. Could you review it again, to check if there is something missing?

Best regards,
Jordi Marias I Parella

**Comment by JordiMarias** [comment]
> Hi, again Riebl! I know haven't been active in this PR for quite a long time, but I'd like to contribute to the project and merge it. I've finally reviewed all the comments again and it seems now that everything is in order. Could you review it again, to check if there is something missing?
> 
> Best regards, Jordi Marias

PD: I'm working quickly to make all UT run fine!

**Comment by riebl** [comment]
Hi @JordiMarias, welcome back! :-)
Let me know if you face trouble fixing the unit tests. Otherwise, I will simply wait for your update.

**Comment by JordiMarias** [comment]
I'm having a hard time fixing the remaining failing tests. Most of the problems that appear regard the boost::variant implementation from the fact that there is compatibility with both versions. More specifically, it boils down to the Certificate provider, when is called the "own_certificate()" method it returns the new "CertificateVariant" type of object which is a visitable variable that can either be the v1.2.1 security cert or the v1.3.1 security cert. Apparently, nothing gets returned when there is returned a boost::variant type of message.
I'm working on it but if you can take a glance at it, it would be great!

**Comment by riebl** [comment]
Update: I am currently extracting the security v1.2.1 encoding to the `vanetza::security::v2` namespace, so `vanetza::security` contains only generic code working for both versions. As part of that, I am also refactoring the interfaces slightly to make the version agnostic. That should help us to get your changes into place as soon as I am done with it.

**Comment by JordiMarias** [comment]
Hi Riebl, what are your plans to be able to merge this pull request? Did you were able to extract the V2 security to the vanetza::security::v2? I can try to do it myself in this PR.

**Comment by riebl** [comment]
Hi @JordiMarias, I have been able to extract most of the V2 security code to its namespace. Unfortunately, I have not yet been able to run all tests successfully again.

**Comment by kenog** [comment]
Hi together,
are there any news about this? We would be very interested in using this soon. Let me know if I can help somehow.

Best regards
Keno

**Comment by riebl** [comment]
Hi @kenog,

no news, I had no time yet to complete this effort. However, I could tidy up my branch a little bit and share it with you. Are you interested?

**Comment by kenog** [comment]
Yes, that would be great! Thank you very much :)

**Comment by riebl** [comment]
@kenog I have pushed branch *sec_v2_namespace*. Please note that the SecurityEntity unit test segfaults. It is still in a rough state…

**Comment by riebl** [comment]
Hi @serserHR, this PR is no longer compatible with our upstream *master* branch. Secured v3 messages can be received meanwhile. Support for outgoing v3 secured messages is a top priority but my spare time is quite limited.

**Comment by riebl** [comment]
I consider closing this PR because (most) of its features have been integrated into the master branch in some form.
Any objections?

**Pull request closed without merging**